File: services/tracking-service/internal/domain/eta.go
package domain

import (
	"time"

	"github.com/google/uuid"
)

const (
	TargetPickup   = "pickup"
	TargetDelivery = "delivery"

	ETASourceProviderETA    = "provider_eta"
	ETASourceCarrierETA     = "carrier_eta"
	ETASourceDriverETA      = "driver_eta"
	ETASourceManualOperator = "manual_operator"
	ETASourceCalculated     = "calculated"

	ETAStatusUnavailable = "unavailable"
	ETAStatusAvailable   = "available"
	ETAStatusStale       = "stale"
	ETAStatusExpired     = "expired"
	ETAStatusCompleted   = "completed"

	ETAFreshnessUnknown = "unknown"
	ETAFreshnessFresh   = "fresh"
	ETAFreshnessStale   = "stale"
	ETAFreshnessExpired = "expired"

	ETAQualityUnknown  = "unknown"
	ETAQualityGood     = "good"
	ETAQualityDegraded = "degraded"
	ETAQualityPoor     = "poor"

	ArrivalEarly    = "early"
	ArrivalOnTime   = "on_time"
	ArrivalAtRisk   = "at_risk"
	ArrivalLate     = "late"
	ArrivalUnknown  = "unknown"

	MaxETAHistoryLimit = 200

	TransitionETABecameAvailable = "eta_became_available"
	TransitionETABecameStale     = "eta_became_stale"
	TransitionETAExpired         = "eta_expired"
	TransitionETARestored        = "eta_restored"
	TransitionETACompleted       = "eta_completed"
	TransitionETASourceChanged   = "eta_source_changed"
)

type ETAObservation struct {
	ID                  uuid.UUID
	TenantID            uuid.UUID
	ShipmentID          uuid.UUID
	TargetType          string
	TargetReference     *string
	EstimatedArrivalAt  time.Time
	SourceType          string
	ProviderCode        *string
	ProviderEventID     *string
	DedupKey            string
	SourceObservedAt    time.Time
	ReceivedAt          time.Time
	QualityStatus       string
	QualityReasons      []string
	ProviderConfidence  *float64
	CreatedAt           time.Time
}

type ShipmentETAState struct {
	TenantID           uuid.UUID
	ShipmentID         uuid.UUID
	TargetType         string
	Status             string
	EstimatedArrivalAt *time.Time
	SourceType         *string
	ProviderCode       *string
	SourceObservedAt   *time.Time
	ReceivedAt         *time.Time
	FreshnessStatus    string
	QualityStatus      string
	AgeSeconds         *int64
	DeliveryLagSeconds *int64
	Version            int64
	UpdatedAt          time.Time
}

type ETATargetSummary struct {
	Status                  string
	EstimatedArrivalAt      *time.Time
	SourceType              *string
	Provider                *string
	SourceObservedAt        *time.Time
	ReceivedAt              *time.Time
	AgeSeconds              *int64
	FreshnessStatus         string
	QualityStatus           string
	QualityReasons          []string
	ProviderConfidence      *float64
	DeliveryLagSeconds      *int64
	PlannedArrivalAt        *time.Time
	ProjectedDeviationSeconds *int64
	ArrivalProjection       string
}

type ShipmentETASummary struct {
	ShipmentID uuid.UUID
	Delivery   *ETATargetSummary
	Pickup     *ETATargetSummary
}

type ETAStateTransition struct {
	ID             uuid.UUID
	TenantID       uuid.UUID
	ShipmentID     uuid.UUID
	TargetType     string
	TransitionType string
	FromStatus     *string
	ToStatus       string
	Metadata       map[string]any
	OccurredAt     time.Time
}

func SourcePriority(sourceType string) int {
	switch sourceType {
	case ETASourceProviderETA:
		return 100
	case ETASourceCarrierETA:
		return 90
	case ETASourceDriverETA:
		return 80
	case ETASourceManualOperator:
		return 70
	case ETASourceCalculated:
		return 60
	default:
		return 0
	}
}

func IsEnabledSourceType(sourceType string) bool {
	switch sourceType {
	case ETASourceProviderETA, ETASourceCarrierETA, ETASourceDriverETA, ETASourceManualOperator:
		return true
	default:
		return false
	}
}

File: apps/web-admin/types/eta.ts
export type ETATargetStatus = 'unavailable' | 'available' | 'stale' | 'expired' | 'completed'

export type ETAFreshnessStatus = 'unknown' | 'fresh' | 'stale' | 'expired'

export type ETAQualityStatus = 'unknown' | 'good' | 'degraded' | 'poor'

export type ETAArrivalProjection = 'early' | 'on_time' | 'at_risk' | 'late' | 'unknown'

export type ETASourceType =
  | 'provider_eta'
  | 'carrier_eta'
  | 'driver_eta'
  | 'manual_operator'
  | 'calculated'

export interface ETATargetSummary {
  status: ETATargetStatus
  estimatedArrivalAt?: string | null
  sourceType?: ETASourceType | null
  provider?: string | null
  sourceObservedAt?: string | null
  receivedAt?: string | null
  ageSeconds?: number | null
  freshnessStatus: ETAFreshnessStatus
  qualityStatus: ETAQualityStatus
  qualityReasons?: string[]
  providerConfidence?: number | null
  deliveryLagSeconds?: number | null
  plannedArrivalAt?: string | null
  projectedDeviationSeconds?: number | null
  arrivalProjection: ETAArrivalProjection
}

export interface ShipmentETASummary {
  shipmentId: string
  delivery?: ETATargetSummary | null
  pickup?: ETATargetSummary | null
}

export interface ETAObservation {
  id: string
  shipmentId: string
  targetType: string
  estimatedArrivalAt: string
  sourceType: ETASourceType
  provider?: string | null
  providerEventId?: string | null
  sourceObservedAt: string
  receivedAt: string
  qualityStatus: ETAQualityStatus
  qualityReasons?: string[]
  providerConfidence?: number | null
}

export interface ETAHistoryResponse {
  items: ETAObservation[]
  total: number
  limit: number
  offset: number
}

export interface ShipmentETAContext {
  plannedPickupAt?: string | null
  plannedDeliveryAt?: string | null
  actualPickupAt?: string | null
  actualDeliveryAt?: string | null
  shipmentStatus?: string | null
}
